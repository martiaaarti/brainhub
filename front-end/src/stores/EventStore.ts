import axios from "axios";
import { DateTime } from "luxon";
import {
  types,
  Instance,
  SnapshotIn,
  flow,
  getSnapshot,
  applySnapshot,
} from "mobx-state-tree";

import { LuxonDateTimeOrNull } from "./types";

export const EventStore = types
  .model("EventStore", {
    firstName: types.optional(types.string, ""),
    lastName: types.optional(types.string, ""),
    email: types.optional(types.string, ""),
    date: types.maybeNull(LuxonDateTimeOrNull),
  })
  .actions((self) => ({
    setFirstName(newFirstName: string) {
      self.firstName = newFirstName;
    },
    setLastName(newLastName: string) {
      self.lastName = newLastName;
    },
    setEmail(newEmail: string) {
      self.email = newEmail;
    },
    setDate(newDate: DateTime | null) {
      self.date = newDate;
    },
  }))
  .actions((self) => {
    let initialState = {};
    return {
      afterCreate: () => {
        initialState = getSnapshot(self);
      },
      reset: () => {
        applySnapshot(self, initialState);
      },
    };
  })
  .actions((self) => ({
    createEvent: flow(function* createEvent() {
      const data = {
        firstName: self.firstName,
        lastName: self.lastName,
        email: self.email,
        date: self.date,
      };
      yield axios({
        method: "post",
        url: "/events",
        baseURL: process.env.REACT_APP_BASE_URL,
        data,
      });
    }),
  }));

export interface EventStoreInstance extends Instance<typeof EventStore> {}
export interface EventStoreSnapshot extends SnapshotIn<typeof EventStore> {}
