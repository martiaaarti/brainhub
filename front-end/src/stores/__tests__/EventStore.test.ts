import { DateTime } from "luxon";
import nock from "nock";
import * as R from "ramda";

import { EventStore, EventStoreInstance } from "../EventStore";

const { REACT_APP_BASE_URL } = process.env;

const eventFixture = {
  lastName: "Kowalska",
  email: "test123@gmail.com",
  firstName: "Grażyna",
  date: "2021-06-01T00:00:00.000+02:00",
};

describe("EventStore", () => {
  it("can create an instance", () => {
    const store = EventStore.create({
      email: "test123@gmail.com",
      firstName: "Grażyna",
      lastName: "Nowak",
      date: null,
    });

    expect(store).toMatchSnapshot();
  });
  describe("actions", () => {
    let store: EventStoreInstance;
    beforeEach(() => {
      store = EventStore.create();
    });

    it("first name can be set", () => {
      expect(store.firstName).toBe("");
      store.setFirstName("Grażyna");
      expect(store.firstName).toBe("Grażyna");
    });

    it("last name can be set", () => {
      expect(store.lastName).toBe("");
      store.setLastName("Nowak");
      expect(store.lastName).toBe("Nowak");
    });

    it("email can be set", () => {
      expect(store.email).toBe("");
      store.setEmail("test123@gmail.com");
      expect(store.email).toBe("test123@gmail.com");
    });

    it("date can be set", () => {
      const date = DateTime.local(2021, 6, 1);
      store.setDate(date);
      expect(store.date).toMatchSnapshot();
    });
  });

  describe("api actions", () => {
    it("can succesfully create an event", async () => {
      const store = EventStore.create(eventFixture);

      store.setLastName("Kowalska");

      const mockPostRequest = nock(
        REACT_APP_BASE_URL || "http://localhost:5000"
      )
        .post("/events", (body) => R.equals(body, eventFixture))
        .reply(
          200,
          {},
          {
            "Access-Control-Allow-Origin": "*",
          }
        );

      await store.createEvent();
      expect(mockPostRequest.isDone()).toBe(true);
    });
  });
});
