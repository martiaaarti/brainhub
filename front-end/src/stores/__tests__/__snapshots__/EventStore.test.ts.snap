// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventStore actions date can be set 1`] = `"2021-06-01T00:00:00.000+02:00"`;

exports[`EventStore can create an instance 1`] = `
Object {
  "date": null,
  "email": "test123@gmail.com",
  "firstName": "Gra≈ºyna",
  "lastName": "Nowak",
}
`;
